modules/build/src/main/scala/scala/build/bsp/BspImpl.scala

```diff
@@ -135,40 +143,35 @@ final class BspImpl(
       buildChangedTest
     )
 
-    (mainScope, testScope)
+    // ala a a
+    PreBuildProject(mainScope, testScope, persistentLogger.diagnostics)
   }
 
   private def buildE(
     actualLocalServer: BspServer,
     bloopServer: BloopServer,
     notifyChanges: Boolean
   ): Either[(BuildException, Scope), Unit] = either {
-    val (preBuildDataMain, preBuildDataTest) =
-      value(prepareBuild(actualLocalServer))
-    if (notifyChanges && (preBuildDataMain.buildChanged || preBuildDataTest.buildChanged))
+    val preBuild = value(prepareBuild(actualLocalServer))
+
+    if (notifyChanges && (preBuild.mainScope.buildChanged || preBuild.testScope.buildChanged))
       notifyBuildChange(actualLocalServer)
-    Build.buildOnce(
-      inputs,
-      preBuildDataMain.sources,
-      inputs.generatedSrcRoot(Scope.Main),
-      preBuildDataMain.generatedSources,
-      preBuildDataMain.buildOptions,
-      Scope.Main,
-      logger,
-      actualLocalClient,
-      bloopServer
-    ).swap.map(e => (e, Scope.Main)).swap
-    Build.buildOnce(
-      inputs,
-      preBuildDataTest.sources,
-      inputs.generatedSrcRoot(Scope.Test),
-      preBuildDataTest.generatedSources,
-      preBuildDataTest.buildOptions,
-      Scope.Test,
-      logger,
-      actualLocalClient,
-      bloopServer
-    ).swap.map(e => (e, Scope.Test)).swap
+
+    def doBuildOnce(data: PreBuildData, scope: Scope) =
+      Build.buildOnce(
+        inputs,
+        data.sources,
+        inputs.generatedSrcRoot(scope),
+        data.generatedSources,
+        data.buildOptions,
+        scope,
+        logger,
+        actualLocalClient,
+        bloopServer
+      ).left.map(_ -> scope)
+
+    value(doBuildOnce(preBuild.mainScope, Scope.Main))
```

Comment:
Please use for-comprehensions as long as we have problems with left-types mismatch in EitherCps

Reply:
sure